text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1796 @@
+
+For example, we are permitted to transmute a `Bool` into a [`u8`]:
+```rust
+#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(u8)]
+enum Bool {
+    True = 1,
+    False = 0,
+}
+
+let _ : u8 = Bool::True.transmute_into();
+let _ : u8 = Bool::False.transmute_into();
+```
+<sup>
+(Note: <code>#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]</code> annotation connotes that <i>all</i> aspects of <code>Bool</code>'s layout are part of its <a href="#-when-is-a-transmutation-stable">library stability guarantee</a>.)
+</sup>
+
+...because all possible instances of `Bool` are also valid instances of [`u8`]. However, transmuting a [`u8`] into a `Bool` is forbidden:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _ : Bool = u8::default().transmute_into(); // Compile Error!
+```
+...because not all instances of [`u8`] are valid instances of `Bool`.
+
+Another example: While laying out certain types, Rust may insert padding bytes between the layouts of fields. In the below example `Padded` has two padding bytes, while `Packed` has none:
+```rust
+#[repr(C)]
+#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
+struct Padded(pub u8, pub u16, pub u8);
+
+#[repr(C)]
+#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
+struct Packed(pub u16, pub u16, pub u16);
+
+assert_eq!(mem::size_of::<Packed>(), mem::size_of::<Padded>());
+```
+
+We may safely transmute from `Packed` to `Padded`:
+```rust
+let _ : Padded = Packed::default().transmute_into();
+```
+...but not from `Padded` to `Packed`:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _ : Packed = Padded::default().transmute_into(); // Compile Error!
+```
+...because doing so would expose two uninitialized padding bytes in `Padded` as if they were initialized bytes in `Packed`.
+
+##### Preserve or Shrink Size
+[owned-size]: #Preserve-or-Shrink-Size
+
+It's completely sound to transmute into a type with fewer bytes than the source type; e.g.:
+```rust
+let _ : [u8; 16] = [u8; 32]::default().transmute_into();
+```
+This transmute truncates away the final sixteen bytes of the `[u8; 32]` value.
+
+A value may ***not*** be transmuted into a type of greater size, if doing so would expose uninitialized bytes as initialized:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _ : [u8; 32] = [u8; 16]::default().transmute_into(); // Compile Error!
+```
+
+#### Requirements on References
+[transmute-references]: #requirements-on-references
+
+The [restrictions above that apply to transmuting owned values][transmute-owned] also apply to transmuting references. However, references carry a few additional restrictions.
+
+A [sound transmutation] must:
+ - [preserve or shrink size][reference-size],
+ - [preserve or relax alignment][reference-alignment],
+ - [preserve or shrink lifetimes][reference-lifetimes],
```

Comment:
Shouldn't this also include mutability? That is, it is unsound to transmute from a `&T` to a `&mut T`. 

Also, these constraints must be followed for sound transmutations if there are references inside of the source or destination types, right? Should that be more explicitly mentioned?

Reply:
> Shouldn't this also include mutability? That is, it is unsound to transmute from a `&T` to a `&mut T`.

This is addressed by the heading ["Preserve or Shrink Uniqueness"](https://github.com/jswrenn/rfcs/blob/safer-transmute/text/0000-safer-transmute.md#preserve-or-shrink-uniqueness). 

> Also, these constraints must be followed for sound transmutations if there are references inside of the source or destination types, right? Should that be more explicitly mentioned?

That's the gist, yes! I'm in the process of creating a [Redex](https://docs.racket-lang.org/redex/) model of this section's semantics for use as an implementation reference.